### ISSUE TITLE
Unable to inject RestTemplateBuilder when Sleuth and TestRestTemplate are used

### ISSUE THREAD

**Issue by fabriziocucci** [spring-cloud/spring-cloud-sleuth#877]
### Affected Version
* 1.3.2.RELEASE

### Description

Apologies in advance if this isn't really an issue but a wrong configuration on my side.

I have a simple spring boot app with the following configuration:

```java
@Configuration
public class MyConfiguration {

    @Bean
    public RestTemplate restTemplate(RestTemplateBuilder restTemplateBuilder) {
        return restTemplateBuilder.build();
    }

    // ...

}
```

A simple test in which I'm injecting a `TestRestTemplate` instance:

```java
@RunWith(SpringRunner.class)
@SpringBootTest(webEnvironment=WebEnvironment.RANDOM_PORT)
public class MyTest {

    @Autowired
    private TestRestTemplate testRestTemplate;
    
    // ...

}
```

If I run the build, this unfortunately fails with the error:

```
***************************
APPLICATION FAILED TO START
***************************

Description:

The dependencies of some of the beans in the application context form a cycle:

   org.springframework.boot.test.web.client.TestRestTemplate
┌─────┐
|  restTemplateBuilder defined in class path resource [org/springframework/boot/autoconfigure/web/WebClientAutoConfiguration$RestTemplateConfiguration.class]
↑     ↓
|  org.springframework.cloud.sleuth.instrument.web.client.TraceWebClientAutoConfiguration$TraceInterceptorConfiguration (field private java.util.Collection org.springframework.cloud.sleuth.instrument.web.client.TraceWebClientAutoConfiguration$TraceInterceptorConfiguration.restTemplates)
↑     ↓
|  restTemplate defined in class path resource [com/company/MyConfiguration.class]
└─────┘
```

But if I do **not** inject the `RestTemplateBuilder` in my configuration:

```java
@Configuration
public class MyConfiguration {

    @Bean
    public RestTemplate restTemplate() {
        return new RestTemplateBuilder().build();
    }

    // ...

}
```

everything works fine!

Is it something expected? If yes, would it be worth documenting it?

Please also note that this issue only appears when adding Sleuth to the application.

**Comment by marcingrzejszczak** [comment]
Can you check the latest snapshots and see if the problem is still there?

**Comment by fabriziocucci** [comment]
Same issue with `1.3.3.BUILD-SNAPSHOT` unfortunately!

**Comment by marcingrzejszczak** [comment]
I can't replicate the problem. 

> Same issue with 1.3.3.BUILD-SNAPSHOT unfortunately!

Are you sure?

```
The dependencies of some of the beans in the application context form a cycle:

   org.springframework.boot.test.web.client.TestRestTemplate
┌─────┐
|  restTemplateBuilder defined in class path resource [org/springframework/boot/autoconfigure/web/WebClientAutoConfiguration$RestTemplateConfiguration.class]
↑     ↓
|  org.springframework.cloud.sleuth.instrument.web.client.TraceWebClientAutoConfiguration$TraceInterceptorConfiguration (field private java.util.Collection org.springframework.cloud.sleuth.instrument.web.client.TraceWebClientAutoConfiguration$TraceInterceptorConfiguration.restTemplates)
↑     ↓
|  restTemplate defined in class path resource [com/company/MyConfiguration.class]
```

There's no such field as 

```
TraceInterceptorConfiguration (field private java.util.Collection org.springframework.cloud.sleuth.instrument.web.client.TraceWebClientAutoConfiguration$TraceInterceptorConfiguration.restTemplates)
```

There's no rest template collection injection here (https://github.com/spring-cloud/spring-cloud-sleuth/blob/1.3.x/spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/TraceWebClientAutoConfiguration.java) anymore. Can you upload a sample that can replicate the problem?

**Comment by fabriziocucci** [comment]
Apologies, when I tested this with the latest snapshot, I've updated my `gradle.build` with:

```
compile('org.springframework.cloud:spring-cloud-starter-sleuth:1.3.3.BUILD-SNAPSHOT')
```

I didn't realize the corresponding `spring-cloud-sleuth-core` dependency was not brought in.

I've now updated the `build.gradle` as follows:

```
compile('org.springframework.cloud:spring-cloud-starter-sleuth:1.3.3.BUILD-SNAPSHOT')
compile('org.springframework.cloud:spring-cloud-sleuth-core:1.3.3.BUILD-SNAPSHOT')
```

and I can't replicate the issue (as you correctly pointed out @marcingrzejszczak).

In summary, this seems to be an issue for version `1.3.2.RELEASE` only.

---

OT: is that the correct way for testing against a shapshot version using Gradle?
